### DISCUSSION TITLE
Неправильное распознавание координат нажатия на рисовалке

### DISCUSSION THREAD

**Discussion by DKarnakov** [vvip-68/LedPanelWiFi#105] [Вопросы и ответы] [answered]
При использовании режима рисования в браузере телефона неправильно распознаются координаты нажатий. Нажимаю на один квадратик, а точка ставится на пару квадратов выше. Причём разница смещения зависит от размера матрицы. Чем матрица меньше, тем смещение больше. Можно это как-то «пофиксить»? 
И ещё один неудобный момент. Раньше, когда матрица управлялась через Android приложение, можно было рисовать, например, линии, если пальцем провести по экрану. Сейчас же ставится только точка, и перемещения пальца по экрану не отслеживаются. Тоже буду благодарен за фикс этого момента. Уж очень дети любят рисовать на гирлянде.

**Comment by vvip-68** [comment]
Позиционирование. Это, видимо, проблемы вашего браузера или телефона. На трех моих телефонах я никакого смещения не замечал. То что рисуется и во что тыкается - просто страничка в браузере. А другие странички в браузере - навигация по ссылкам у вас работает? Нет смещений?

По поводу рисования сплошной линий - наверное возможно, но требует несколько более углубленного программирования нежели тычки. Вероятно, когда-нибудь я этим займусь и возможно сделаю, но, думается, не в ближайшее предновогоднее время. Вы же на следующий год планируете пользоваться гирляндой с этим проектом? Или у вас это просто доска для рисования, используемая круглый год?

**Reply by DKarnakov** [reply]
Да, пользуемся только на новогодние праздники. Обновлю тогда прошивку к следующему НГ. Дети пока с ноутбука рисовать будут. Там нет таких проблем с позиционированием.

**Reply by vvip-68** [reply]
Что лишний раз, как бы, подтверждает - проблема не веб-приложения, а телефона. Веб-приложение то одно и то же работает - что на телефоне, что в браузере ноутбука. А размер матрицы вообще какой?

**Reply by DKarnakov** [reply]
А по поводу браузера, то возможно вы правы. Я пользовался на iPhone, хотя и с разными браузерами. Везде такая ерунда со смещением. Хотя все другие контролы работают исправно — кнопки жмутся, селекторы переключаются. Может это проблема именно этого фрейма рисования.
На стационарных компьютерах и ноутбуке таких проблем нет. 
Возможно и на Андроиде тоже, но мне проверить негде.

**Reply by vvip-68** [reply]
Ну вот, а у меня айфона нет. Это не фрейм - это HTML-элемент <canvas> в плоской структуре странички безо всяких многоуровностей и вложенностей. Какие координаты тычок в него пальцем он сообщает - те и передаются в программу. Если нет проблем на андроиде и на компе - скорее всего проблема в браузере айфона...

Хотя вот на айфонах же на ретине плотность пикселей картинки в два раза больше чем в стандартных экранах - может это как-то влияет. Но тогда наверное ошибка бы накапливалась в зависимости от того в какой квадратик тыкаешь. Например в левый верхний - попадало бы без ошибок, в середине -, скажем, с ошибкой на 1-2, а в правом нижнем углу - уже еще больше или вообще бы не реагировало... 

 Но опять же мне даже потренироваться не на чем. Я не могу исправлять "ошибки" которые не могу воспроизвести.

**Reply by DKarnakov** [reply]
> А размер матрицы вообще какой?

18 ширина, 11 высота.
198 диодов

**Reply by DKarnakov** [reply]
И да, вы оказались правы. Нашел старый планшет на Андроиде, проверил там - все нажатия отрабатываются корректно. Значит это проблема iOS. Причем любого браузера (пробовал Яндекс, Chrome, Safari)

**Reply by vvip-68** [reply]
Если есть дух авантюризма или запасной микроконтроллер - можете обновить прошивку. В последней версии есть рисование пальцем - двигая его по полю вместо тычков в клеточки.

Если запасного контроллера нет... Смотрите, вдруг что-нибудь пойдет не так и прям перед НГ вообще ничего работать не будет...

**Reply by DKarnakov** [reply]
Получилось! Даже на iPhone рисует линии без смещения. А вот рисовалка «тычками» стала вести себя по-другому. Теперь она реагирует на 2 события - нажатие и отпускание. Так вот, при нажатии точка ставится где надо, но при отпускании ставится вторая точка со смещением.
Я записал [видео](https://disk.yandex.ru/i/mVEQLH3OsBveaw) с телефона. Посмотрите, может будут идеи почему так может быть.

**Reply by vvip-68** [reply, marked as answer]
Так, ну iPhone у меня нет - проверить работоспособность скрипта не могу.
В браузере на компе - эмуляция айфона работает без вот таких вот артефактов.

Давайте так - я чуть-чуть поменял скрипт рисования - проверьте, может в этой версии починится.
Нужно скачать последнюю версию с репозитория.
Скетч проекта компилировать / загружать не нужно.
Нужно просто залить обновленные скрипты Web-интерфейса из папки data,
перезапустить браузер / обновить страницу, чтобы айфон загрузил обновленные скрипты.
Проверить есть ли артефакты при тычках.

**Reply by DKarnakov** [reply]
Залил обновление web интерфейса. Теперь все работает как надо! Никаких смещений и ненужных артефактов. Нажатия отрабатываются как и должны.
Спасибо вам большое.
С наступающим!
